File: python/ui.py
import pyxel as px
import time
import math

import PyxelUniversalFont as puf

from constants import *
from utils import estimate_text_width, calculate_text_center_position
from managers import LanguageManager

class Notification:
    def __init__(self, message, duration=NOTIFICATION_MAX_DISPLAY_TIME, msg_type="info",
                 wrap_text_func=None, max_wrap_width=0):
        self.message = message
        self.start_time = time.time()
        self.duration = duration
        self.msg_type = msg_type
        self.is_alive = True

        self.state = "fading_in"
        self.current_x = None
        self.current_y = None
        self.target_x = None
        self.target_y = None
        self.vel_x = 0

        self._wrap_text = wrap_text_func
        self._max_wrap_width = max_wrap_width
        self._box_width = None
        self._box_height = None

    def get_text_color(self):
        if self.msg_type == "error":
            return NOTIFICATION_TEXT_COLOR_ERROR
        elif self.msg_type == "success":
            return NOTIFICATION_TEXT_COLOR_SUCCESS
        return NOTIFICATION_TEXT_COLOR_INFO

    def _calculate_dimensions(self):
        if not self._wrap_text:
            self._box_width = 0
            self._box_height = 0
            return

        lines = self._wrap_text(self.message, self._max_wrap_width)
        total_text_height = len(lines) * FONT_SIZE + (max(0, len(lines) - 1)) * NOTIFICATION_LINE_SPACING
        self._box_height = total_text_height + NOTIFICATION_PADDING_Y * 2

        max_line_width = max(estimate_text_width(line) for line in lines) if lines else 0
        self._box_width = min(NOTIFICATION_MAX_WIDTH, max_line_width + NOTIFICATION_PADDING_X * 2)

    def set_target_position(self, target_x, target_y):
        self.target_x = target_x
        self.target_y = target_y

        if self.current_y is None:
            self.current_x = self.target_x
            self.current_y = self.target_y + NOTIFICATION_FADE_IN_OFFSET_Y

    def update(self):
        if not self.is_alive:
            return

        if self._box_width is None:
            self._calculate_dimensions()
            if self._box_width == 0 or self._box_height == 0:
                self.is_alive = False
                return

        if self.current_y is not None and self.target_y is not None and abs(self.target_y - self.current_y) > NOTIFICATION_TARGET_Y_TOLERANCE:
            move_amount_y = NOTIFICATION_FADE_IN_AMOUNT_Y_PER_FRAME
            if self.current_y < self.target_y:
                self.current_y += min(move_amount_y, self.target_y - self.current_y)
            else:
                self.current_y -= min(move_amount_y, self.current_y - self.target_y)
        elif self.target_y is not None:
            self.current_y = self.target_y

        if self.state == "fading_in":
            if self.current_y is not None and self.target_y is not None and abs(self.target_y - self.current_y) <= NOTIFICATION_TARGET_Y_TOLERANCE:
                self.state = "visible"
        elif self.state == "visible":
            if time.time() - self.start_time > self.duration:
                self.state = "fading_out"
                self.vel_x = NOTIFICATION_FADE_OUT_INITIAL_AMOUNT_X_PER_FRAME
        elif self.state == "fading_out":
            self.vel_x += NOTIFICATION_FADE_OUT_ACCELERATION_X_PER_FRAME
            self.current_x += self.vel_x
            if self.current_x > SCREEN_WIDTH:
                self.is_alive = False

    def get_draw_position(self):
        return (self.current_x, self.current_y) if self.current_x is not None and self.current_y is not None else None

    def get_box_dimensions(self):
        return (self._box_width, self._box_height) if self._box_width is not None and self._box_height is not None else None

    def get_wrapped_lines(self):
        return self._wrap_text(self.message, self._max_wrap_width) if self._wrap_text and self._max_wrap_width > 0 else [self.message]

class SelectBlock:
    def __init__(self):
        self._selection_effect_start_time = 0
        self._is_effect_currently_active = False

    def update_selection_status(self, is_mouse_over_a_block):
        if is_mouse_over_a_block and not self._is_effect_currently_active:
            self._is_effect_currently_active = True
            self._selection_effect_start_time = time.time()
        elif not is_mouse_over_a_block:
            self._is_effect_currently_active = False

    def draw(self, mouse_x, mouse_y, is_game_active_and_not_menu):
        if not is_game_active_and_not_menu or not self._is_effect_currently_active:
            return

        grid_aligned_x = math.floor(mouse_x / BLOCK_SIZE) * BLOCK_SIZE
        grid_aligned_y = math.floor(mouse_y / BLOCK_SIZE) * BLOCK_SIZE
        elapsed_time = time.time() - self._selection_effect_start_time

        if elapsed_time <= 1.0:
            px.blt(grid_aligned_x, grid_aligned_y, *SPRITE_SELECT_NORMAL)
        elif elapsed_time <= 2.0:
            px.blt(grid_aligned_x - 1, grid_aligned_y - 1, *SPRITE_SELECT_LARGE)
        else:
            self._selection_effect_start_time = time.time()

class ButtonBox:
    def __init__(self, font_writer: puf.Writer, lang_manager: LanguageManager):
        self.font = font_writer
        self.lang_manager = lang_manager

    def _get_mouse_status_on_button(self, x, y, w, h):
        is_hover = x <= px.mouse_x <= x + w - 1 and y <= px.mouse_y <= y + h - 1
        is_pressed = is_hover and px.btn(px.MOUSE_BUTTON_LEFT)
        is_released = is_hover and px.btnr(px.MOUSE_BUTTON_LEFT)
        return is_pressed, is_hover, is_released

    def draw_button(self, x, y, w, h, text_key='text', pressed_text_key='press'):
        is_pressed, _, is_released = self._get_mouse_status_on_button(x, y, w, h)
        current_text_str = self.lang_manager.get_string(pressed_text_key if is_pressed else text_key)
        bg_color = COLOR_BUTTON_PRESSED_BG if is_pressed else COLOR_BUTTON_BG
        px.rect(x, y, w, h, bg_color)
        px.rectb(x, y, w, h, COLOR_BUTTON_BORDER)
        if not is_pressed:
            px.line(x + w - 1, y + 1, x + w - 1, y + h - 2, 0)
            px.line(x + 1, y + h - 1, x + w - 2, y + h - 1, 0)
        text_x_offset, text_y_offset = calculate_text_center_position(w, h, current_text_str)
        self.font.draw(x + text_x_offset, y + text_y_offset, current_text_str, FONT_SIZE, COLOR_BUTTON_TEXT)
        return is_released

class GameMenu:
    def __init__(self, game_instance_ref, lang_manager_ref: LanguageManager, font_writer_ref: puf.Writer):
        self.game = game_instance_ref
        self.lang_manager = lang_manager_ref
        self.font = font_writer_ref
        self.button_handler = ButtonBox(self.font, self.lang_manager)
        self.width = 100
        self.height = SCREEN_HEIGHT - (SCREEN_HEIGHT // 10 * 2) + MENU_ITEM_HEIGHT // 2 * 3
        self.x = (SCREEN_WIDTH - self.width) // 2
        self.y = (SCREEN_HEIGHT - self.height) // 2
        self.menu_items_def = [
            {"key": "menu.sound_effects", "type": "checkbox", "setting_attr": "se_on"},
            {"key": "menu.music", "type": "checkbox", "setting_attr": "bgm_on"},
            {"key": "menu.language", "type": "dropdown"},
            {"key": "button.menu.save.default", "press_key":"button.menu.save.pressed", "type": "button", "action_label": "Save Game"},
            {"key": "button.menu.load.default", "press_key":"button.menu.load.pressed", "type": "button", "action_label": "Load Game"},
            {"key": "button.menu.quit.default", "press_key":"button.menu.quit.pressed", "type": "button", "action_label": "Quit Game"}
        ]
        self.selected_button_action = None
        self.is_lang_dropdown_open = False
        self.lang_dropdown_trigger_rect = None
        self.lang_dropdown_options_rects = []

    def _draw_checkbox(self, x, y, label_text, is_checked):
        check_y = y + (MENU_ITEM_HEIGHT - CHECKBOX_SIZE) // 2
        px.rectb(x, check_y, CHECKBOX_SIZE, CHECKBOX_SIZE, 0)
        if is_checked:
            px.rect(x + 2, check_y + 2, CHECKBOX_SIZE - 4, CHECKBOX_SIZE - 4, 0)
        self.font.draw(x + CHECKBOX_SIZE + CHECKBOX_TEXT_GAP, y + (MENU_ITEM_HEIGHT - FONT_SIZE) // 2, label_text, FONT_SIZE, 0)

    def handle_input(self):
        if not self.game.is_menu_visible:
            self.is_lang_dropdown_open = False
            return

        clicked_outside_dropdown = False
        if self.is_lang_dropdown_open and px.btnp(px.MOUSE_BUTTON_LEFT):
            clicked_on_option = False
            for opt_x, opt_y, opt_w, opt_h, lang_code in self.lang_dropdown_options_rects:
                if opt_x <= px.mouse_x < opt_x + opt_w and opt_y <= px.mouse_y < opt_y + opt_h:
                    self.lang_manager.set_language(lang_code)
                    self.game.current_language_code = lang_code
                    self.game.update_window_title()
                    self.is_lang_dropdown_open = False
                    clicked_on_option = True
                    break
            if not clicked_on_option and self.lang_dropdown_trigger_rect and not (self.lang_dropdown_trigger_rect[0] <= px.mouse_x < self.lang_dropdown_trigger_rect[0] + self.lang_dropdown_trigger_rect[2] and self.lang_dropdown_trigger_rect[1] <= px.mouse_y < self.lang_dropdown_trigger_rect[1] + self.lang_dropdown_trigger_rect[3]):
                clicked_outside_dropdown = True

        if px.btnp(px.MOUSE_BUTTON_LEFT) and (not self.is_lang_dropdown_open or clicked_outside_dropdown):
            current_y = self.y + MENU_ITEM_HEIGHT + MENU_PADDING * 2
            for item_def in self.menu_items_def:
                item_x = self.x + MENU_PADDING
                item_w = self.width - MENU_PADDING * 2
                item_rect = (item_x, current_y, item_w, MENU_ITEM_HEIGHT)
                if item_rect[0] <= px.mouse_x < item_rect[0] + item_rect[2] and item_rect[1] <= px.mouse_y < item_rect[1] + item_rect[3]:
                    if item_def["type"] == "checkbox":
                        setattr(self.game, item_def["setting_attr"], not getattr(self.game, item_def["setting_attr"]))
                        break
                    elif item_def["type"] == "dropdown":
                        self.is_lang_dropdown_open = not self.is_lang_dropdown_open
                        break
                current_y += MENU_ITEM_HEIGHT + MENU_PADDING

        if clicked_outside_dropdown:
            self.is_lang_dropdown_open = False

    def draw(self):
        if not self.game.is_menu_visible:
            self.is_lang_dropdown_open = False
            return None

        self.selected_button_action = None
        px.rect(self.x, self.y, self.width, self.height, 9)
        px.rectb(self.x, self.y, self.width, self.height, 0)
        menu_title_str = self.lang_manager.get_string("menu.title")
        title_x_offset, title_y_offset = calculate_text_center_position(self.width, MENU_ITEM_HEIGHT, menu_title_str)
        self.font.draw(self.x + title_x_offset, self.y + MENU_PADDING + title_y_offset, menu_title_str, FONT_SIZE, 0)

        current_y = self.y + MENU_ITEM_HEIGHT + MENU_PADDING * 2
        dropdown_draw_params = None

        for item_def in self.menu_items_def:
            item_x = self.x + MENU_PADDING
            item_w = self.width - MENU_PADDING * 2
            if item_def["type"] == "checkbox":
                self._draw_checkbox(item_x, current_y, self.lang_manager.get_string(item_def["key"]), getattr(self.game, item_def["setting_attr"]))
            elif item_def["type"] == "dropdown":
                lang_label_prefix = self.lang_manager.get_string(item_def["key"]) + ":"
                self.font.draw(item_x, current_y + (MENU_ITEM_HEIGHT - FONT_SIZE) // 2, lang_label_prefix, FONT_SIZE, 0)
                dropdown_button_x = item_x + estimate_text_width(lang_label_prefix) + MENU_PADDING
                dropdown_button_w = item_w - (estimate_text_width(lang_label_prefix) + MENU_PADDING)
                current_lang_display_name = self.lang_manager.get_available_languages().get(self.lang_manager.current_lang_code, self.lang_manager.current_lang_code)
                px.rectb(dropdown_button_x, current_y, dropdown_button_w, MENU_ITEM_HEIGHT, 0)
                self.font.draw(dropdown_button_x + MENU_PADDING, current_y + (MENU_ITEM_HEIGHT - FONT_SIZE) // 2, current_lang_display_name, FONT_SIZE, 0)
                px.tri(dropdown_button_x + dropdown_button_w - DROPDOWN_ARROW_WIDTH - 2, current_y + (MENU_ITEM_HEIGHT - DROPDOWN_ARROW_HEIGHT)//2, dropdown_button_x + dropdown_button_w - 2 - DROPDOWN_ARROW_WIDTH//2, current_y + (MENU_ITEM_HEIGHT + DROPDOWN_ARROW_HEIGHT)//2, dropdown_button_x + dropdown_button_w - 2, current_y + (MENU_ITEM_HEIGHT - DROPDOWN_ARROW_HEIGHT)//2, 0)
                self.lang_dropdown_trigger_rect = (dropdown_button_x, current_y, dropdown_button_w, MENU_ITEM_HEIGHT)
                if self.is_lang_dropdown_open:
                    dropdown_draw_params = (dropdown_button_x, current_y, dropdown_button_w)
            elif item_def["type"] == "button":
                if self.button_handler.draw_button(item_x, current_y, item_w, MENU_ITEM_HEIGHT, item_def["key"], item_def["press_key"]):
                    if not self.is_lang_dropdown_open:
                        self.selected_button_action = item_def["action_label"]
            current_y += MENU_ITEM_HEIGHT + MENU_PADDING

        if dropdown_draw_params:
            dd_x, dd_y, dd_w = dropdown_draw_params
            options_y_start = dd_y + MENU_ITEM_HEIGHT
            available_langs = self.lang_manager.get_available_languages()
            max_option_width = dd_w
            for _, display_name_opt in available_langs.items():
                 max_option_width = max(max_option_width, estimate_text_width(display_name_opt) + MENU_PADDING * 2)
            list_bg_x, list_bg_y, list_bg_w, list_bg_h = dd_x, options_y_start, max_option_width, len(available_langs) * MENU_ITEM_HEIGHT
            px.rect(list_bg_x, list_bg_y, list_bg_w, list_bg_h, 13)
            px.rectb(list_bg_x, list_bg_y, list_bg_w, list_bg_h, 0)
            self.lang_dropdown_options_rects = []
            current_opt_y = options_y_start
            for lang_code, display_name in available_langs.items():
                option_rect_for_draw = (list_bg_x, current_opt_y, list_bg_w, MENU_ITEM_HEIGHT)
                self.lang_dropdown_options_rects.append((*option_rect_for_draw, lang_code))
                if option_rect_for_draw[0] <= px.mouse_x < option_rect_for_draw[0] + option_rect_for_draw[2] and option_rect_for_draw[1] <= px.mouse_y < option_rect_for_draw[1] + option_rect_for_draw[3]:
                    px.rect(*option_rect_for_draw, 1)
                self.font.draw(list_bg_x + MENU_PADDING, current_opt_y + (MENU_ITEM_HEIGHT - FONT_SIZE) // 2, display_name, FONT_SIZE, 0)
                current_opt_y += MENU_ITEM_HEIGHT
        return self.selected_button_action

File: src/ui.rs
use crate::constants::*;
use crate::render::sprites::*;
use macroquad::prelude::*;

#[derive(PartialEq)]
pub enum NotificationState {
    FadingIn,
    Visible,
    FadingOut,
}

pub struct Notification {
    pub message: String,
    pub duration: f64,
    pub msg_type: String, // info, error, success
    pub is_alive: bool,
    pub state: NotificationState,

    start_time: f64,
    current_x: f32,
    current_y: f32,
    target_x: f32,
    target_y: f32,
    vel_x: f32,

    // Computed layout
    box_width: f32,
    box_height: f32,
    wrapped_lines: Vec<String>,
}

impl Notification {
    pub fn new(
        message: String,
        duration: f64,
        msg_type: &str,
        max_width: f32,
        font: Option<&Font>,
    ) -> Self {
        let mut n = Self {
            message: message.clone(),
            duration,
            msg_type: msg_type.to_string(),
            is_alive: true,
            state: NotificationState::FadingIn,
            start_time: get_time(),
            current_x: 0.0,
            current_y: 0.0,
            target_x: 0.0,
            target_y: 0.0,
            vel_x: 0.0,
            box_width: 0.0,
            box_height: 0.0,
            wrapped_lines: Vec::new(),
        };
        n.calculate_dimensions(max_width, font);
        n
    }

    fn calculate_dimensions(&mut self, max_width: f32, font: Option<&Font>) {
        let font = match font {
            Some(f) => f,
            None => return, // Cannot calculate without a font
        };

        let measure = |s: &str| -> f32 { measure_text(s, Some(font), FONT_SIZE as u16, 1.0).width };

        let space_width = measure(" ");
        let words: Vec<&str> = self.message.split(' ').collect();
        let mut lines = Vec::new();
        let mut current_line = String::new();
        let mut current_line_width = 0.0;

        for word in words {
            let word_width = measure(word);

            if word_width > max_width {
                if !current_line.is_empty() {
                    lines.push(current_line);
                    current_line = String::new();
                    current_line_width = 0.0;
                }

                let mut temp_word_line = String::new();
                let mut temp_word_width = 0.0;
                for char in word.chars() {
                    let char_str = char.to_string();
                    let char_width = measure(&char_str);
                    if temp_word_width + char_width > max_width {
                        lines.push(temp_word_line);
                        temp_word_line = char_str;
                        temp_word_width = char_width;
                    } else {
                        temp_word_line.push(char);
                        temp_word_width += char_width;
                    }
                }
                if !temp_word_line.is_empty() {
                    current_line = temp_word_line;
                    current_line_width = temp_word_width;
                }
            } else {
                let width_if_added = if current_line.is_empty() {
                    word_width
                } else {
                    current_line_width + space_width + word_width
                };

                if width_if_added > max_width && !current_line.is_empty() {
                    lines.push(current_line);
                    current_line = word.to_string();
                    current_line_width = word_width;
                } else {
                    if !current_line.is_empty() {
                        current_line.push(' ');
                    }
                    current_line.push_str(word);
                    current_line_width = measure(&current_line);
                }
            }
        }
        if !current_line.is_empty() {
            lines.push(current_line);
        }

        self.wrapped_lines = lines;

        let line_count = self.wrapped_lines.len();
        self.box_height = (line_count as f32 * FONT_SIZE)
            + (line_count.saturating_sub(1) as f32 * NOTIFICATION_LINE_SPACING)
            + NOTIFICATION_PADDING_Y * 2.0;

        let max_line_w = self
            .wrapped_lines
            .iter()
            .map(|l| measure(l))
            .fold(0.0f32, f32::max);

        self.box_width = (max_line_w + NOTIFICATION_PADDING_X * 2.0).min(NOTIFICATION_MAX_WIDTH);
    }

    pub fn set_target_position(&mut self, x: f32, y: f32) {
        self.target_x = x;
        self.target_y = y;
        // Init position if new
        if self.current_x == 0.0 && self.current_y == 0.0 {
            self.current_x = x;
            self.current_y = y + NOTIFICATION_FADE_IN_OFFSET_Y;
        }
    }

    pub fn update(&mut self) {
        if !self.is_alive {
            return;
        }

        // Y Movement
        let dy = self.target_y - self.current_y;
        if dy.abs() > NOTIFICATION_TARGET_Y_TOLERANCE {
            let move_dist = NOTIFICATION_FADE_IN_AMOUNT_Y_PER_FRAME.min(dy.abs());
            self.current_y += if dy > 0.0 { move_dist } else { -move_dist };
        } else {
            self.current_y = self.target_y;
        }

        // State Machine
        match self.state {
            NotificationState::FadingIn => {
                if (self.current_y - self.target_y).abs() <= NOTIFICATION_TARGET_Y_TOLERANCE {
                    self.state = NotificationState::Visible;
                }
            }
            NotificationState::Visible => {
                if get_time() - self.start_time > self.duration {
                    self.state = NotificationState::FadingOut;
                    self.vel_x = NOTIFICATION_FADE_OUT_INITIAL_AMOUNT_X_PER_FRAME;
                }
            }
            NotificationState::FadingOut => {
                self.vel_x += NOTIFICATION_FADE_OUT_ACCELERATION_X_PER_FRAME;
                self.current_x += self.vel_x;
                if self.current_x > SCREEN_WIDTH {
                    self.is_alive = false;
                }
            }
        }
    }

    pub fn draw_high_res(&self, font: Option<&Font>, scale: f32, off_x: f32, off_y: f32) {
        if !self.is_alive {
            return;
        }

        let sx = off_x + self.current_x * scale;
        let sy = off_y + self.current_y * scale;
        let sw = self.box_width * scale;
        let sh = self.box_height * scale;

        draw_rectangle(sx.floor(), sy.floor(), sw, sh, NOTIFICATION_BG_COLOR);

        let text_col = if self.msg_type == "error" {
            NOTIFICATION_TEXT_COLOR_ERROR
        } else if self.msg_type == "success" {
            NOTIFICATION_TEXT_COLOR_SUCCESS
        } else {
            NOTIFICATION_TEXT_COLOR_INFO
        };

        draw_rectangle_lines(sx.floor(), sy.floor(), sw, sh, 1.0, text_col);

        let mut y_off = sy + NOTIFICATION_PADDING_Y * scale;
        let s_font_size = (FONT_SIZE * scale).floor() as u16;
        for line in &self.wrapped_lines {
            draw_text_ex(
                line,
                (sx + NOTIFICATION_PADDING_X * scale).floor(),
                (y_off + FONT_SIZE * scale * 0.8).floor(),
                TextParams {
                    font_size: s_font_size,
                    font: font.or_else(|| TextParams::default().font),
                    color: text_col,
                    ..Default::default()
                },
            );
            y_off += (FONT_SIZE + NOTIFICATION_LINE_SPACING) * scale;
        }
    }

    pub fn get_box_width(&self) -> f32 {
        self.box_width
    }
    pub fn get_box_height(&self) -> f32 {
        self.box_height
    }
}

pub struct SelectBlock {
    selection_effect_start_time: f64,
    is_effect_active: bool,
    block_coords: Option<(f32, f32)>,
    preview_sprite: Option<Rect>,
    is_valid: bool,
}

impl SelectBlock {
    pub fn new() -> Self {
        Self {
            selection_effect_start_time: 0.0,
            is_effect_active: false,
            block_coords: None,
            preview_sprite: None,
            is_valid: true,
        }
    }

    pub fn update(
        &mut self,
        hovered_block_coords: Option<(f32, f32)>,
        preview_sprite: Option<Rect>,
        is_valid: bool,
    ) {
        if let Some(coords) = hovered_block_coords {
            if !self.is_effect_active {
                self.selection_effect_start_time = get_time();
            } else if self.block_coords != Some(coords) {
                self.selection_effect_start_time = get_time();
            }
            self.is_effect_active = true;
            self.block_coords = Some(coords);
        } else {
            self.is_effect_active = false;
            self.block_coords = None;
        }
        self.preview_sprite = preview_sprite;
        self.is_valid = is_valid;
    }

    pub fn draw(
        &mut self,
        camera_x: f32,
        camera_y: f32,
        atlas: &Texture2D,
    ) {
        if !self.is_effect_active {
            return;
        }

        let (world_block_x, world_block_y) = self.block_coords.unwrap();

        let screen_x = (world_block_x - camera_x).round();
        let screen_y = (world_block_y - camera_y).round();
        let elapsed = get_time() - self.selection_effect_start_time;

        if let Some(sprite) = self.preview_sprite {
            let color = if self.is_valid {
                Color::new(1.0, 1.0, 1.0, 0.4)
            } else {
                Color::new(1.0, 0.3, 0.3, 0.5)
            };
            draw_texture_ex(
                atlas,
                screen_x,
                screen_y,
                color,
                DrawTextureParams {
                    source: Some(sprite),
                    ..Default::default()
                },
            );
        }

        if elapsed > SELECTION_PULSE_DURATION {
            self.selection_effect_start_time = get_time();
        }

        if elapsed <= (SELECTION_PULSE_DURATION / 2.0) {
            draw_texture_ex(
                atlas,
                screen_x,
                screen_y,
                WHITE,
                DrawTextureParams {
                    source: Some(SPRITE_SELECT_NORMAL),
                    ..Default::default()
                },
            );
        } else {
            draw_texture_ex(
                atlas,
                screen_x - SELECTION_ENLARGE_AMOUNT,
                screen_y - SELECTION_ENLARGE_AMOUNT,
                WHITE,
                DrawTextureParams {
                    source: Some(SPRITE_SELECT_LARGE),
                    dest_size: Some(vec2(
                        BLOCK_SIZE + SELECTION_ENLARGE_AMOUNT * 2.0,
                        BLOCK_SIZE + SELECTION_ENLARGE_AMOUNT * 2.0,
                    )),
                    ..Default::default()
                },
            );
        }
    }
}
